File: aermet/core/stations.py
"""
Station discovery — find nearby surface (GHCNh / ISD) and upper air (IGRA)
stations.

Downloads and caches station inventories from NCEI, searches by distance
from a project location, and ranks candidates.

Primary surface station source: GHCNh station list (next-gen replacement
for ISD).  Falls back to ISD history if GHCNh is unavailable.
"""

import csv
import os

from .http import get as http_get

from .geocode import haversine_km
from .igra import parse_igra_station_list

_DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')

# GHCNh (preferred — the actual data source we download from)
GHCNH_STATION_LIST_URL = "https://www.ncei.noaa.gov/oa/global-historical-climatology-network/hourly/doc/ghcnh-station-list.txt"
GHCNH_INVENTORY_URL = "https://www.ncei.noaa.gov/oa/global-historical-climatology-network/hourly/doc/ghcnh-inventory.txt"

# ISD (legacy fallback)
ISD_HISTORY_URL = "https://www.ncei.noaa.gov/pub/data/noaa/isd-history.csv"

IGRA_STATION_LIST_URL = "https://www.ncei.noaa.gov/data/integrated-global-radiosonde-archive/doc/igra2-station-list.txt"


# ---------------------------------------------------------------------------
# GHCNh surface stations (preferred)
# ---------------------------------------------------------------------------

def ensure_ghcnh_station_list():
    """Download GHCNh station list if not cached locally."""
    path = os.path.join(_DATA_DIR, 'ghcnh-station-list.txt')
    if os.path.exists(path):
        return path
    os.makedirs(_DATA_DIR, exist_ok=True)
    resp = http_get(GHCNH_STATION_LIST_URL, timeout=60)
    resp.raise_for_status()
    with open(path, 'w', encoding='utf-8') as f:
        f.write(resp.text)
    return path


def ensure_ghcnh_inventory():
    """Download GHCNh inventory (obs counts per month) if not cached."""
    path = os.path.join(_DATA_DIR, 'ghcnh-inventory.txt')
    if os.path.exists(path):
        return path
    os.makedirs(_DATA_DIR, exist_ok=True)
    resp = http_get(GHCNH_INVENTORY_URL, timeout=120)
    resp.raise_for_status()
    with open(path, 'w', encoding='utf-8') as f:
        f.write(resp.text)
    return path


def load_ghcnh_stations(path=None):
    """
    Parse the GHCNh station list (fixed-width format).

    Format (from NCEI docs):
      Col  1-11: ID (11-char GHCNh station ID)
      Col 13-20: LATITUDE
      Col 22-30: LONGITUDE
      Col 32-37: ELEVATION (m)
      Col 39-40: STATE (US only)
      Col 42-71: NAME
      Col 73-75: GSN FLAG
      Col 77-79: HCN/CRN FLAG
      Col 81-85: WMO ID
      Col 87-90: ICAO
    """
    if path is None:
        path = ensure_ghcnh_station_list()

    stations = []
    with open(path, 'r', encoding='utf-8') as f:
        for line in f:
            if len(line.strip()) < 40:
                continue
            try:
                sid = line[0:11].strip()
                lat = float(line[12:20].strip())
                lon = float(line[21:30].strip())
                elev_str = line[31:37].strip()
                elev = float(elev_str) if elev_str and elev_str != '-999.9' else None
                state = line[38:40].strip() if len(line) > 40 else ''
                name = line[41:71].strip() if len(line) > 71 else ''
                wmo_id = line[80:85].strip() if len(line) > 85 else ''
                icao = line[86:90].strip() if len(line) > 90 else ''

                # Country code is the first 2 chars of the station ID
                country = sid[:2] if len(sid) >= 2 else ''

                stations.append({
                    'ghcnh_id': sid,
                    'lat': lat,
                    'lon': lon,
                    'elev': elev,
                    'state': state,
                    'name': name,
                    'country': country,
                    'wmo_id': wmo_id,
                    'icao': icao,
                })
            except (ValueError, IndexError):
                continue
    return stations


def load_ghcnh_inventory(path=None):
    """
    Parse the GHCNh inventory file to get observation counts per station/month.

    Returns dict: { station_id: { (year, month): count, ... }, ... }

    The inventory is a fixed-width file with format similar to GHCNd:
      Col 1-11:  ID
      Col 13-16: YEAR
      Col 18-19: MONTH
      Col 21+:   COUNT (number of records)
    """
    if path is None:
        path = ensure_ghcnh_inventory()

    inventory = {}
    with open(path, 'r', encoding='utf-8') as f:
        for line in f:
            if len(line.strip()) < 20:
                continue
            try:
                sid = line[0:11].strip()
                year = int(line[12:16].strip())
                month = int(line[17:19].strip())
                count = int(line[20:].strip())
                if sid not in inventory:
                    inventory[sid] = {}
                inventory[sid][(year, month)] = count
            except (ValueError, IndexError):
                continue
    return inventory


def get_ghcnh_station_year_range(inventory, station_id):
    """Get first and last year of data for a GHCNh station from inventory."""
    months = inventory.get(station_id, {})
    if not months:
        return None, None
    years = [ym[0] for ym in months.keys()]
    return min(years), max(years)


def check_ghcnh_completeness(inventory, station_id, start_year, end_year,
                              min_obs_per_month=200):
    """
    Quick data completeness screen for a GHCNh station.

    Returns dict with:
      - complete_months: count of months meeting min_obs_per_month
      - total_months: total months in requested range
      - pct_complete: percentage
    """
    months = inventory.get(station_id, {})
    total = 0
    complete = 0
    for year in range(start_year, end_year + 1):
        for month in range(1, 13):
            total += 1
            count = months.get((year, month), 0)
            if count >= min_obs_per_month:
                complete += 1
    pct = round(complete / total * 100, 1) if total > 0 else 0
    return {
        'complete_months': complete,
        'total_months': total,
        'pct_complete': pct,
    }


def find_nearby_ghcnh_stations(lat, lon, start_year=None, end_year=None,
                                max_distance_km=150, max_results=20):
    """
    Find GHCNh surface stations near a location.

    If start/end year provided and inventory is available, includes
    a basic completeness check.  Returns list sorted by distance.
    """
    all_stations = load_ghcnh_stations()

    # Try to load inventory for completeness screening
    inv = None
    if start_year and end_year:
        try:
            inv = load_ghcnh_inventory()
        except Exception:
            pass  # inventory is optional

    candidates = []
    for s in all_stations:
        dist = haversine_km(lat, lon, s['lat'], s['lon'])
        if dist > max_distance_km:
            continue

        s_copy = dict(s)
        s_copy['distance_km'] = round(dist, 1)

        if inv and start_year and end_year:
            first, last = get_ghcnh_station_year_range(inv, s['ghcnh_id'])
            if first is not None:
                s_copy['first_year'] = first
                s_copy['last_year'] = last
                # Skip stations with no data overlap
                if last < start_year or first > end_year:
                    continue
                comp = check_ghcnh_completeness(
                    inv, s['ghcnh_id'], start_year, end_year)
                s_copy['completeness'] = comp
            else:
                # No inventory data — include but flag
                s_copy['first_year'] = None
                s_copy['last_year'] = None

        candidates.append(s_copy)

    candidates.sort(key=lambda x: x['distance_km'])
    return candidates[:max_results]


# ---------------------------------------------------------------------------
# ISD surface stations (legacy fallback)
# ---------------------------------------------------------------------------

def ensure_isd_station_list():
    """Download ISD station history CSV if not cached locally."""
    path = os.path.join(_DATA_DIR, 'isd-history.csv')
    if os.path.exists(path):
        return path
    os.makedirs(_DATA_DIR, exist_ok=True)
    resp = http_get(ISD_HISTORY_URL, timeout=60)
    resp.raise_for_status()
    with open(path, 'w', encoding='utf-8') as f:
        f.write(resp.text)
    return path


def load_isd_stations(path=None):
    """Parse ISD station history CSV into list of dicts."""
    if path is None:
        path = ensure_isd_station_list()

    stations = []
    with open(path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            try:
                lat = row.get('LAT', '').strip()
                lon = row.get('LON', '').strip()
                if not lat or not lon or lat == '' or lon == '':
                    continue
                lat = float(lat)
                lon = float(lon)
                if lat == 0 and lon == 0:
                    continue

                usaf = row.get('USAF', '999999').strip()
                wban = row.get('WBAN', '99999').strip()
                name = row.get('STATION NAME', '').strip()
                ctry = row.get('CTRY', '').strip()
                state = row.get('STATE', '').strip()
                icao = row.get('ICAO', '').strip()
                elev = row.get('ELEV(M)', '').strip()
                begin = row.get('BEGIN', '').strip()
                end = row.get('END', '').strip()

                # Parse begin/end dates (YYYYMMDD format)
                first_year = int(begin[:4]) if begin and len(begin) >= 4 else 9999
                last_year = int(end[:4]) if end and len(end) >= 4 else 0

                stations.append({
                    'usaf': usaf,
                    'wban': wban,
                    'name': name,
                    'country': ctry,
                    'state': state,
                    'icao': icao,
                    'lat': lat,
                    'lon': lon,
                    'elev': float(elev) if elev else None,
                    'first_year': first_year,
                    'last_year': last_year,
                    'begin': begin,
                    'end': end,
                })
            except (ValueError, KeyError):
                continue
    return stations


def find_nearby_surface_stations(lat, lon, start_year, end_year,
                                  max_distance_km=150, max_results=20):
    """
    Find surface stations near a location that have data coverage
    for the requested period.  Tries GHCNh first (preferred), falls
    back to ISD if GHCNh station list isn't available.
    Returns list sorted by distance.
    """
    try:
        return find_nearby_ghcnh_stations(
            lat, lon, start_year, end_year,
            max_distance_km=max_distance_km, max_results=max_results)
    except Exception:
        pass

    # Fallback to ISD
    all_stations = load_isd_stations()
    candidates = []

    for s in all_stations:
        dist = haversine_km(lat, lon, s['lat'], s['lon'])
        if dist > max_distance_km:
            continue
        # Check date range overlap
        if s['last_year'] < start_year or s['first_year'] > end_year:
            continue
        s_copy = dict(s)
        s_copy['distance_km'] = round(dist, 1)
        candidates.append(s_copy)

    candidates.sort(key=lambda x: x['distance_km'])
    return candidates[:max_results]


# ---------------------------------------------------------------------------
# IGRA upper air stations
# ---------------------------------------------------------------------------

def ensure_igra_station_list():
    """Download IGRA station list if not cached locally."""
    path = os.path.join(_DATA_DIR, 'igra2-station-list.txt')
    if os.path.exists(path):
        return path
    os.makedirs(_DATA_DIR, exist_ok=True)
    resp = http_get(IGRA_STATION_LIST_URL, timeout=60)
    resp.raise_for_status()
    with open(path, 'w', encoding='utf-8') as f:
        f.write(resp.text)
    return path


def load_igra_stations(path=None):
    """Load IGRA station inventory."""
    if path is None:
        path = ensure_igra_station_list()
    return parse_igra_station_list(path)


def find_nearby_upper_air_stations(lat, lon, start_year, end_year,
                                    max_distance_km=400, max_results=10):
    """
    Find IGRA upper air stations near a location with data coverage.
    Uses a larger search radius than surface since UA stations are sparser.
    """
    all_stations = load_igra_stations()
    candidates = []

    for s in all_stations:
        dist = haversine_km(lat, lon, s['lat'], s['lon'])
        if dist > max_distance_km:
            continue
        if s['last_year'] < start_year or s['first_year'] > end_year:
            continue
        s_copy = dict(s)
        s_copy['distance_km'] = round(dist, 1)
        candidates.append(s_copy)

    candidates.sort(key=lambda x: x['distance_km'])
    return candidates[:max_results]

File: aermet-tool/src/stations.rs
/// Station discovery — find nearby surface (GHCNh/ISD) and upper air (IGRA) stations.
use crate::geocode::haversine_km;
use crate::http_client::client;
use crate::igra;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

const GHCNH_STATION_LIST_URL: &str =
    "https://www.ncei.noaa.gov/oa/global-historical-climatology-network/hourly/doc/ghcnh-station-list.txt";
const GHCNH_INVENTORY_URL: &str =
    "https://www.ncei.noaa.gov/oa/global-historical-climatology-network/hourly/doc/ghcnh-inventory.txt";
const ISD_HISTORY_URL: &str = "https://www.ncei.noaa.gov/pub/data/noaa/isd-history.csv";

fn data_dir() -> PathBuf {
    let exe_dir = std::env::current_exe()
        .ok()
        .and_then(|p| p.parent().map(|p| p.to_path_buf()))
        .unwrap_or_else(|| PathBuf::from("."));
    exe_dir.join("data")
}

// ---------------------------------------------------------------------------
// GHCNh surface stations
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GhcnhStation {
    pub ghcnh_id: String,
    pub lat: f64,
    pub lon: f64,
    pub elev: Option<f64>,
    #[serde(default)]
    pub state: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub country: String,
    #[serde(default)]
    pub wmo_id: String,
    #[serde(default)]
    pub icao: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub distance_km: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub first_year: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_year: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completeness: Option<CompletenessInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompletenessInfo {
    pub complete_months: i32,
    pub total_months: i32,
    pub pct_complete: f64,
}

async fn ensure_file(url: &str, filename: &str, timeout_secs: u64) -> Result<PathBuf, String> {
    let dir = data_dir();
    fs::create_dir_all(&dir).ok();
    let path = dir.join(filename);
    if path.exists() {
        return Ok(path);
    }
    let resp = client()
        .get(url)
        .timeout(std::time::Duration::from_secs(timeout_secs))
        .send()
        .await
        .map_err(|e| format!("Download {filename}: {e}"))?;
    let text = resp.text().await.map_err(|e| format!("Read {filename}: {e}"))?;
    fs::write(&path, &text).map_err(|e| format!("Write {filename}: {e}"))?;
    Ok(path)
}

pub async fn ensure_ghcnh_station_list() -> Result<PathBuf, String> {
    ensure_file(GHCNH_STATION_LIST_URL, "ghcnh-station-list.txt", 60).await
}

pub async fn ensure_ghcnh_inventory() -> Result<PathBuf, String> {
    ensure_file(GHCNH_INVENTORY_URL, "ghcnh-inventory.txt", 120).await
}

pub async fn ensure_isd_station_list() -> Result<PathBuf, String> {
    ensure_file(ISD_HISTORY_URL, "isd-history.csv", 60).await
}

pub async fn ensure_igra_station_list() -> Result<PathBuf, String> {
    ensure_file(igra::IGRA_STATION_LIST_URL, "igra2-station-list.txt", 60).await
}

fn load_ghcnh_stations(text: &str) -> Vec<GhcnhStation> {
    let mut stations = Vec::new();
    for line in text.lines() {
        if line.trim().len() < 40 {
            continue;
        }
        let sid = line.get(0..11).unwrap_or("").trim().to_string();
        let lat = line.get(12..20).and_then(|s| s.trim().parse::<f64>().ok());
        let lon = line.get(21..30).and_then(|s| s.trim().parse::<f64>().ok());
        let elev_str = line.get(31..37).unwrap_or("").trim();
        let elev = if elev_str.is_empty() || elev_str == "-999.9" {
            None
        } else {
            elev_str.parse::<f64>().ok()
        };
        let state = line.get(38..40).unwrap_or("").trim().to_string();
        let name = line.get(41..71).unwrap_or("").trim().to_string();
        let wmo_id = line.get(80..85).unwrap_or("").trim().to_string();
        let icao = line.get(86..90).unwrap_or("").trim().to_string();
        let country = if sid.len() >= 2 {
            sid[..2].to_string()
        } else {
            String::new()
        };

        if let (Some(lat), Some(lon)) = (lat, lon) {
            stations.push(GhcnhStation {
                ghcnh_id: sid,
                lat,
                lon,
                elev,
                state,
                name,
                country,
                wmo_id,
                icao,
                distance_km: None,
                first_year: None,
                last_year: None,
                completeness: None,
            });
        }
    }
    stations
}

type Inventory = HashMap<String, HashMap<(i32, i32), i32>>;

fn load_ghcnh_inventory(text: &str) -> Inventory {
    let mut inventory: Inventory = HashMap::new();
    for line in text.lines() {
        if line.trim().len() < 20 {
            continue;
        }
        let sid = line.get(0..11).unwrap_or("").trim();
        let year = line.get(12..16).and_then(|s| s.trim().parse::<i32>().ok());
        let month = line.get(17..19).and_then(|s| s.trim().parse::<i32>().ok());
        let count = line.get(20..).and_then(|s| s.trim().parse::<i32>().ok());
        if let (Some(y), Some(m), Some(c)) = (year, month, count) {
            inventory
                .entry(sid.to_string())
                .or_default()
                .insert((y, m), c);
        }
    }
    inventory
}

fn check_ghcnh_completeness(
    inv: &Inventory,
    station_id: &str,
    start_year: i32,
    end_year: i32,
    min_obs: i32,
) -> CompletenessInfo {
    let months = inv.get(station_id);
    let mut total = 0;
    let mut complete = 0;
    for year in start_year..=end_year {
        for month in 1..=12 {
            total += 1;
            let count = months
                .and_then(|m| m.get(&(year, month)))
                .copied()
                .unwrap_or(0);
            if count >= min_obs {
                complete += 1;
            }
        }
    }
    let pct = if total > 0 {
        (complete as f64 / total as f64 * 100.0 * 10.0).round() / 10.0
    } else {
        0.0
    };
    CompletenessInfo {
        complete_months: complete,
        total_months: total,
        pct_complete: pct,
    }
}

fn year_range_from_inventory(inv: &Inventory, station_id: &str) -> (Option<i32>, Option<i32>) {
    match inv.get(station_id) {
        Some(months) => {
            let years: Vec<i32> = months.keys().map(|(y, _)| *y).collect();
            (years.iter().copied().min(), years.iter().copied().max())
        }
        None => (None, None),
    }
}

pub async fn find_nearby_surface_stations(
    lat: f64,
    lon: f64,
    start_year: i32,
    end_year: i32,
    max_distance_km: f64,
) -> Result<Vec<serde_json::Value>, String> {
    // Try GHCNh first
    match try_ghcnh_stations(lat, lon, start_year, end_year, max_distance_km).await {
        Ok(stations) => return Ok(stations),
        Err(_) => {}
    }

    // Fallback to ISD
    find_isd_stations(lat, lon, start_year, end_year, max_distance_km).await
}

async fn try_ghcnh_stations(
    lat: f64,
    lon: f64,
    start_year: i32,
    end_year: i32,
    max_distance_km: f64,
) -> Result<Vec<serde_json::Value>, String> {
    let stn_path = ensure_ghcnh_station_list().await?;
    let stn_text = fs::read_to_string(&stn_path).map_err(|e| format!("Read: {e}"))?;
    let all_stations = load_ghcnh_stations(&stn_text);

    // Try loading inventory
    let inv = match ensure_ghcnh_inventory().await {
        Ok(inv_path) => {
            let inv_text = fs::read_to_string(&inv_path).unwrap_or_default();
            Some(load_ghcnh_inventory(&inv_text))
        }
        Err(_) => None,
    };

    let mut candidates = Vec::new();
    for s in &all_stations {
        let dist = haversine_km(lat, lon, s.lat, s.lon);
        if dist > max_distance_km {
            continue;
        }

        let mut sc = s.clone();
        sc.distance_km = Some((dist * 10.0).round() / 10.0);

        if let Some(ref inv) = inv {
            let (first, last) = year_range_from_inventory(inv, &s.ghcnh_id);
            sc.first_year = first;
            sc.last_year = last;
            if let (Some(first), Some(last)) = (first, last) {
                if last < start_year || first > end_year {
                    continue;
                }
                sc.completeness = Some(check_ghcnh_completeness(
                    inv,
                    &s.ghcnh_id,
                    start_year,
                    end_year,
                    200,
                ));
            }
        }

        candidates.push(sc);
    }

    candidates.sort_by(|a, b| {
        a.distance_km
            .unwrap_or(f64::MAX)
            .partial_cmp(&b.distance_km.unwrap_or(f64::MAX))
            .unwrap()
    });
    candidates.truncate(20);

    let json: Vec<serde_json::Value> = candidates
        .iter()
        .map(|s| serde_json::to_value(s).unwrap_or_default())
        .collect();
    Ok(json)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct IsdStation {
    usaf: String,
    wban: String,
    name: String,
    country: String,
    state: String,
    icao: String,
    lat: f64,
    lon: f64,
    elev: Option<f64>,
    first_year: i32,
    last_year: i32,
    begin: String,
    end: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    distance_km: Option<f64>,
}

async fn find_isd_stations(
    lat: f64,
    lon: f64,
    start_year: i32,
    end_year: i32,
    max_distance_km: f64,
) -> Result<Vec<serde_json::Value>, String> {
    let path = ensure_isd_station_list().await?;
    let text = fs::read_to_string(&path).map_err(|e| format!("Read ISD: {e}"))?;
    let mut rdr = csv::Reader::from_reader(text.as_bytes());

    let mut candidates = Vec::new();
    for result in rdr.records() {
        let row = match result {
            Ok(r) => r,
            Err(_) => continue,
        };
        let lat_s = row.get(6).unwrap_or("").trim();
        let lon_s = row.get(7).unwrap_or("").trim();
        if lat_s.is_empty() || lon_s.is_empty() {
            continue;
        }
        let slat: f64 = match lat_s.parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let slon: f64 = match lon_s.parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        if slat == 0.0 && slon == 0.0 {
            continue;
        }

        let dist = haversine_km(lat, lon, slat, slon);
        if dist > max_distance_km {
            continue;
        }

        let begin = row.get(10).unwrap_or("").trim().to_string();
        let end = row.get(11).unwrap_or("").trim().to_string();
        let first_year: i32 = begin
            .get(..4)
            .and_then(|s| s.parse().ok())
            .unwrap_or(9999);
        let last_year: i32 = end.get(..4).and_then(|s| s.parse().ok()).unwrap_or(0);

        if last_year < start_year || first_year > end_year {
            continue;
        }

        let elev_s = row.get(8).unwrap_or("").trim();

        candidates.push(IsdStation {
            usaf: row.get(0).unwrap_or("999999").trim().to_string(),
            wban: row.get(1).unwrap_or("99999").trim().to_string(),
            name: row.get(2).unwrap_or("").trim().to_string(),
            country: row.get(3).unwrap_or("").trim().to_string(),
            state: row.get(4).unwrap_or("").trim().to_string(),
            icao: row.get(5).unwrap_or("").trim().to_string(),
            lat: slat,
            lon: slon,
            elev: if elev_s.is_empty() {
                None
            } else {
                elev_s.parse().ok()
            },
            first_year,
            last_year,
            begin,
            end,
            distance_km: Some((dist * 10.0).round() / 10.0),
        });
    }

    candidates.sort_by(|a, b| {
        a.distance_km
            .unwrap_or(f64::MAX)
            .partial_cmp(&b.distance_km.unwrap_or(f64::MAX))
            .unwrap()
    });
    candidates.truncate(20);

    let json: Vec<serde_json::Value> = candidates
        .iter()
        .map(|s| serde_json::to_value(s).unwrap_or_default())
        .collect();
    Ok(json)
}

pub async fn find_nearby_upper_air_stations(
    lat: f64,
    lon: f64,
    start_year: i32,
    end_year: i32,
    max_distance_km: f64,
) -> Result<Vec<serde_json::Value>, String> {
    let path = ensure_igra_station_list().await?;
    let text = fs::read_to_string(&path).map_err(|e| format!("Read IGRA: {e}"))?;
    let all_stations = igra::parse_igra_station_list(&text);

    let mut candidates: Vec<serde_json::Value> = Vec::new();
    for s in &all_stations {
        let dist = haversine_km(lat, lon, s.lat, s.lon);
        if dist > max_distance_km {
            continue;
        }
        if s.last_year < start_year || s.first_year > end_year {
            continue;
        }
        let mut val = serde_json::to_value(s).unwrap_or_default();
        val["distance_km"] = serde_json::json!(((dist * 10.0).round() / 10.0));
        candidates.push(val);
    }

    candidates.sort_by(|a, b| {
        let da = a["distance_km"].as_f64().unwrap_or(f64::MAX);
        let db = b["distance_km"].as_f64().unwrap_or(f64::MAX);
        da.partial_cmp(&db).unwrap()
    });
    candidates.truncate(10);

    Ok(candidates)
}
